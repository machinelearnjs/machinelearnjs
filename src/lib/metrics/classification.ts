import * as tf from '@tensorflow/tfjs';
import * as _ from 'lodash';
import { Type1DMatrix } from '../types';
import { ValidationClassMismatch, ValidationError } from '../utils/Errors';
import { reshape } from '../utils/tensors';
import { checkArray } from '../utils/validation';

/**
 * util function to calculate a weighted sum
 * @param {any} sampleScore
 * @param {any} normalize
 * @returns {number}
 * @ignore
 */
function _weightedSum({
  sampleScore,
  // sampleWeight = null,
  normalize = false,
}): number {
  if (normalize) {
    return _.mean(sampleScore);
  } else {
    return _.sum(sampleScore);
  }
}

/**
 * Validator for classification exceptions
 * @param y_true
 * @param y_pred
 * @param labels
 * @param options
 * @ignore
 */
export const validateInitialInputs = (
  y_true: Type1DMatrix<string | number>,
  y_pred: Type1DMatrix<string | number>,
  labels,
  options = {},
) => {
  const checkMultiClass = _.get(options, 'multiclass');

  // Multiclass
  if (checkMultiClass) {
    // TODO: Multi label
    if (checkArray(y_true).multiclass || checkArray(y_pred).multiclass) {
      throw new ValidationError('Multiclass is not supported yet!');
    }
  }

  // Checking nullity or empty
  if (!y_true || _.isEmpty(y_true)) {
    throw new ValidationError('y_true cannot be null or empty');
  }
  if (!y_pred || _.isEmpty(y_pred)) {
    throw new ValidationError('y_pred cannot be null or empty');
  }

  // Checking the size equality
  if (_.size(y_true) !== _.size(y_pred)) {
    throw new ValidationError('y_true and y_pred are not equal in size!');
  }

  // Checking labels equal to both y_true and y_pred classes
  // Labels is optional
  if (labels) {
    const yTrueCls = _.flowRight((x: Type1DMatrix<string | number>) => _.sortBy(x, (y) => y), (x) => _.uniq(x))(y_true);

    const yPredCls = _.flowRight((x: Type1DMatrix<string | number>) => _.sortBy(x, (y) => y), (x) => _.uniq(x))(y_pred);

    const sortedLabels = _.sortBy(labels, (x) => x);
    if (!_.isEqual(sortedLabels, yTrueCls) || !_.isEqual(sortedLabels, yPredCls)) {
      throw new ValidationClassMismatch('Labels must match the classes');
    }
  }
};

/**
 * Accuracy classification score.
 *
 * In multilabel classification, this function computes subset accuracy:
 * the set of labels predicted for a sample must exactly match the corresponding set of labels in y_true.
 *
 * @example
 * import { accuracyScore } from 'machinelearn/metrics';
 *
 * const accResult = accuracyScore(
 *  [0, 1, 2, 3],
 *  [0, 2, 1, 3]
 * );
 *
 * // accuracy result: 0.5
 *
 * @param y_true - 1d array-like, or label indicator array / sparse matrix
 * @param y_pred - 1d array-like, or label indicator array / sparse matrix
 * @param normalize
 */
export function accuracyScore(
  y_true: Type1DMatrix<number | string> = null,
  y_pred: Type1DMatrix<number | string> = null,
  {
    normalize = true,
  }: // sample_weight = null
  {
    normalize: boolean;
  } = {
    normalize: true,
  },
): number {
  validateInitialInputs(y_true, y_pred, null, { multiclass: true });

  const yTrueRange = _.range(0, _.size(y_true));
  const normalised = _.map(yTrueRange, (index) => {
    const yTrue = y_true[index];
    const yPred = y_pred[index];
    return yTrue === yPred ? 1 : 0;
  });

  return _weightedSum({
    normalize,
    sampleScore: normalised,
  });
}

/**
 * Zero-one classification loss.
 *
 * If normalize is `true`, return the fraction of misclassifications (float),
 * else it returns the number of misclassifications (int). The best performance is 0.
 *
 * @example
 * import { zeroOneLoss } from 'machinelearn/metrics';
 *
 * const loss_zero_one_result = zeroOneLoss(
 *   [1, 2, 3, 4],
 *   [2, 2, 3, 5]
 * );
 * console.log(loss_zero_one_result); // 0.5
 *
 * @param {any} y_true - Ground truth (correct) labels.
 * @param {any} y_pred - Predicted labels, as returned by a classifier.
 * @param {any} normalize
 * @returns {number}
 */
export function zeroOneLoss(
  y_true = null,
  y_pred = null,
  {
    /**
     * If False, return the number of misclassifications. Otherwise, return the fraction of misclassifications.
     */
    normalize = true,
  }: {
    normalize: boolean;
  } = {
    normalize: true,
  },
): number {
  if (normalize) {
    return 1 - accuracyScore(y_true, y_pred);
  }
  // TODO: Fix return 0; implement when normalize === false
  return 0;
}

/**
 * A confusion matrix is a technique for summarizing the performance of a classification algorithm.
 *
 * Classification accuracy alone can be misleading if you have an unequal number of observations in each class or if you have more than two classes in your dataset.
 *
 * Calculating a confusion matrix can give you a better idea of what your classification model is getting right and what types of errors it is making.
 *
 * @example
 * import { confusion_matrix } from 'machinelearn/metrics';
 *
 * const matrix1 = confusion_matrix([1, 2, 3], [1, 2, 3]);
 * console.log(matrix1); // [ [ 1, 0, 0 ], [ 0, 1, 0 ], [ 0, 0, 1 ] ]
 *
 * const matrix2 = confusion_matrix(
 *   ['cat', 'ant', 'cat', 'cat', 'ant', 'bird'],
 *   ['ant', 'ant', 'cat', 'cat', 'ant', 'cat']
 * );
 * console.log(matrix2); // [ [ 1, 2, 0 ], [ 2, 0, 0 ], [ 0, 1, 0 ] ]
 *
 * @param y_true - Ground truth (correct) target values.
 * @param y_pred - Estimated targets as returned by a classifier.
 * @param labels
 */
export function confusion_matrix(
  y_true: Type1DMatrix<string | number> = null,
  y_pred: Type1DMatrix<string | number> = null,
  {
    /**
     * List of labels to index the matrix. This may be used to reorder or
     * select a subset of labels. If none is given, those that appear
     * at least once in y_true or y_pred are used in sorted order.
     */
    labels = null,
  }: {
    labels?: any[];
  } = {
    labels: null,
  },
): number[] {
  validateInitialInputs(y_true, y_pred, labels);

  // TODO: Sorting if set by options
  // TODO: classes should be based on yTrue
  const yTrueCls = _.uniqBy(y_true, (x) => x);
  const yPredCls = _.uniqBy(y_pred, (x) => x);

  // TODO: Issue was raisen to fix the typing: https://github.com/josdejong/mathjs/issues/1150
  const yTrueSize = _.size(yTrueCls);
  // const placeholder: any = math.zeros(_.size(yTrueCls), _.size(yTrueCls));
  const rawZeros = [...tf.zeros([yTrueSize, yTrueSize]).dataSync()];
  const placeholder: any = reshape(rawZeros, [yTrueSize, yTrueSize]);

  // Calculating the confusion matrix
  // Looping the index for y_true
  const rowRange = _.range(0, _.size(placeholder));
  _.forEach(rowRange, (rowIndex) => {
    // Looping the index for y_pred
    const colRange = _.range(0, _.size(placeholder[rowIndex]));
    _.forEach(colRange, (colIndex) => {
      // Get current target y true and y pred
      const yTargetTrueVal = yTrueCls[rowIndex];
      const yTargetPredVal = yPredCls[colIndex];

      // Looping the range of y true for pairing
      const yTrueRange = _.range(0, _.size(y_true));
      const score = _.reduce(
        yTrueRange,
        (sum, n) => {
          const trueVal = y_true[n];
          const predVal = y_pred[n];

          if (_.isEqual(trueVal, yTargetTrueVal) && _.isEqual(predVal, yTargetPredVal)) {
            return sum + 1;
          }
          return sum;
        },
        0,
      );

      // Recording the score
      placeholder[rowIndex][colIndex] = score;
    });
  });

  return placeholder;
}
