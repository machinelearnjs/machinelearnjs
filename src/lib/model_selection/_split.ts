import * as tf from '@tensorflow/tfjs';
import * as _ from 'lodash';
import { Type1DMatrix, Type2DMatrix } from '../types';
import { ValidationError } from '../utils/Errors';
import RandomState, { RandomStateObj } from '../utils/random';
import { approximateMode, arraySplit, countBin, inferShape, invidualize } from '../utils/tensors';
import { numSamples, validateFitInputs } from '../utils/validation';

const testShapes = (X: Type1DMatrix<any> | Type2DMatrix<any>, y: Type1DMatrix<any>) => {
  const xShape: Type1DMatrix<number> = inferShape(X);
  const yShape: Type1DMatrix<number> = inferShape(y);
  if (xShape.length > 0 && yShape.length > 0 && xShape[0] !== yShape[0]) {
    throw new ValidationError('X and y must have an identical size');
  }
};
/**
 * K-Folds cross-validator
 *
 * Provides train/test indices to split data in train/test sets. Split dataset into k consecutive folds (without shuffling by default).
 *
 * Each fold is then used once as a validation while the k - 1 remaining folds form the training set.
 *
 * @example
 * import { KFold } from 'machinelearn/model_selection';
 *
 * const kFold = new KFold({ k: 5 });
 * const X1 = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1];
 *
 * /* [ { trainIndex: [ 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19 ],
 * *  testIndex: [ 0, 1, 2, 3 ] },
 * * { trainIndex: [ 0, 1, 2, 3, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19 ],
 * *  testIndex: [ 4, 5, 6, 7 ] },
 * * { trainIndex: [ 0, 1, 2, 3, 4, 5, 6, 7, 12, 13, 14, 15, 16, 17, 18, 19 ],
 * *  testIndex: [ 8, 9, 10, 11 ] },
 * * { trainIndex: [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 16, 17, 18, 19 ],
 * *  testIndex: [ 12, 13, 14, 15 ] },
 * * { trainIndex: [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15 ],
 * *  testIndex: [ 16, 17, 18, 19 ] } ]
 *
 */
export class KFold {
  private k: number;
  private shuffle: boolean;

  /**
   *
   * @param {any} k - Number of folds. Must be at least 2.
   * @param {any} shuffle - Whether to shuffle the data before splitting into batches.
   */
  constructor({ k = 2, shuffle = false }) {
    if (k < 2) {
      throw new ValidationError('Number of folds cannot be less than 2');
    }
    this.k = k;
    this.shuffle = shuffle;
  }
  /**
   *
   * @param X - Training data, where n_samples is the number of samples and n_features is the number of features.
   * @param y - The target variable for supervised learning problems.
   * @returns {any[]}
   */
  public split(X: Type1DMatrix<any> = null, y: Type1DMatrix<any> = null): any[] {
    testShapes(X, y);
    if (this.k > X.length || this.k > y.length) {
      throw new ValidationError(
        `Cannot have number of splits k=${this.k} greater than the number of samples: ${_.size(X)}`,
      );
    }

    const binSize = _.floor(_.size(X) / this.k);
    const xRange = _.range(0, _.size(X));
    const splitRange = _.range(0, this.k);
    return _.reduce(
      splitRange,
      (sum, index) => {
        // Calculate binSizeRange according to k value. e.g. 0 -> [0,1]. 1 -> [2, 3].
        const binSizeRange = _.range(index * binSize, index * binSize + binSize);
        // X index range used for test set. It can either be shuffled e.g. [ 2, 0, 1 ] or raw value [ 0, 1, 2 ]
        const testXRange = _.flowRight((x) => (this.shuffle ? _.shuffle(x) : x), () => _.clone(xRange))();
        // Getting testIndex according to binSizeRange from testXRange
        const testIndex = _.reduce(
          binSizeRange,
          (xIndeces, i) => {
            return _.concat(xIndeces, [testXRange[i]]);
          },
          [],
        );
        const trainIndex = _.pullAll(_.clone(xRange), testIndex);
        return _.concat(sum, [{ trainIndex, testIndex }]);
      },
      [],
    );
  }
}

/**
 * Split arrays or matrices into random train and test subsets
 *
 * @example
 * import { train_test_split } from 'machinelearn/model_selection';
 *
 * const X = [[0, 1], [2, 3], [4, 5], [6, 7], [8, 9]];
 * const y = [0, 1, 2, 3, 4];
 *
 * train_test_split(X, y, {
 *   test_size: 0.33,
 *   train_size: 0.67,
 *   random_state: 42
 * });
 *
 * /*
 * * { xTest: [ [ 0, 1 ], [ 8, 9 ] ],
 * *  xTrain: [ [ 4, 5 ], [ 6, 7 ], [ 2, 3 ] ],
 * *  yTest: [ 0, 4 ],
 * *  yTrain: [ 2, 3, 1 ] }
 *
 * @param {any} X - input data
 * @param {any} y - target data
 * @param {number} test_size - size of the returning test set
 * @param {number} train_size - size of the returning training set
 * @param {number} random_state - state used to shuffle data
 * @param {boolean} clone - to clone the original data
 * @returns {{xTest: any[]; xTrain: any[]; yTest: any[]; yTrain: any[]}}
 */
export function train_test_split(
  X: Type2DMatrix<any> = null,
  y: Type1DMatrix<any> = null,
  {
    // Arguments and their default values
    test_size = 0.25,
    train_size = 0.75,
    random_state = 0,
    clone = true,
  }: {
    // Param types
    test_size?: number;
    train_size?: number;
    random_state?: number;
    clone?: boolean;
  } = {
    // Default if nothing is given
    test_size: 0.25,
    train_size: 0.75,
    random_state: 0,
    clone: true,
  },
): {
  xTest: any[];
  xTrain: any[];
  yTest: any[];
  yTrain: any[];
} {
  const _X = clone ? _.cloneDeep(X) : X;
  const _y = clone ? _.cloneDeep(y) : y;
  // Checking if either of these params is not array
  if (!_.isArray(_X) || !_.isArray(_y) || _X.length === 0 || _y.length === 0) {
    throw new ValidationError('X and y must be array and cannot be empty');
  }

  validateFitInputs(_X, _y);
  // Training dataset size accoding to X
  const trainSizeLength: number = _.round(train_size * _X.length);
  const testSizeLength: number = _.round(test_size * _X.length);

  if (_.round(test_size + train_size) !== 1) {
    throw new ValidationError('Sum of test_size and train_size does not equal 1');
  }
  // Initiate Random engine
  const randomEngine: RandomStateObj = new RandomState(random_state);

  // split
  const xTrain = [];
  const yTrain = [];
  const xTest = [];
  const yTest = [];

  // Getting X_train and y_train
  while (xTrain.length < trainSizeLength && yTrain.length < trainSizeLength) {
    const index = randomEngine.rangedInt(0, X.length - 1);

    // X_train
    xTrain.push(_X[index]);
    _X.splice(index, 1);

    // y_train
    yTrain.push(_y[index]);
    _y.splice(index, 1);
  }

  while (xTest.length < testSizeLength) {
    const index = randomEngine.rangedInt(0, _X.length - 1);
    // X test
    xTest.push(_X[index]);
    _X.splice(index, 1);

    // y train
    yTest.push(_y[index]);
    _y.splice(index, 1);
  }

  // Filter return results
  const clean = (items: any[]) => _.filter(items, (item: any) => !_.isUndefined(item));
  return {
    xTest: clean(xTest),
    xTrain: clean(xTrain),
    yTest: clean(yTest),
    yTrain: clean(yTrain),
  };
}

const rangeValidationError = (type, size, n_samples) => `${type}=${size} should be either 
positive and smaller than number of samples ${n_samples} or a float in (0, 1) range`;

const testRangeValidationError = (test_size, n_samples) => rangeValidationError('test_size', test_size, n_samples);

const trainRangeValidationError = (test_size, n_samples) => rangeValidationError('test_size', test_size, n_samples);

/**
 * StratifiedShuffleSplit
 */
export class StratifiedShuffleSplit {
  private n_splits: number;
  private testSize: number;
  private trainSize: number;
  private rng: RandomStateObj;
  private defaultTestSize: number = 0.1;
  constructor(n_splits: number = 10, testSize: number = null, trainSize?: number, seed?: number) {
    this.n_splits = n_splits;
    this.testSize = testSize;
    this.trainSize = trainSize;
    this.rng = new RandomState(seed);
  }

  split = (X: Type1DMatrix<any> | Type2DMatrix<any> = null, y: Type1DMatrix<any> = null): Type2DMatrix<any> => {
    const XTensor = tf.tensor(X);
    const nSamples = numSamples(XTensor);

    const [nTest, nTrain] = validateShuffleSplit(nSamples, this.testSize, this.trainSize, this.defaultTestSize);

    const [classes, yIndices] = invidualize(y);
    const nClasses: number = classes.length;
    const classCounts: Type1DMatrix<number> = countBin(yIndices);
    if (_.min(classCounts) < 2) {
      throw new ValidationError(
        `The least populated class in y=${y} has only 1 member, which is too few. The minimum number of groups for any class cannot be less than 2.`,
      );
    }

    if (nTrain < nClasses) {
      throw new ValidationError(
        `The train_size = ${nTrain} should be greater or equal to the number of classes = ${nClasses}`,
      );
    }

    if (nTest < nClasses) {
      throw new ValidationError(
        `The test_size = ${nTest} should be greater or equal to the number of classes = ${nClasses}`,
      );
    }

    const cumsumClassCounts: tf.Tensor1D = tf.cumsum(classCounts);
    const classIndices = arraySplit(
      yIndices.sort(),
      cumsumClassCounts.slice(0, cumsumClassCounts.shape[0] - 1).arraySync(),
    );

    const test = [];
    const train = [];
    for (let i = 0; i < this.n_splits; i++) {
      const n_i: Type1DMatrix<number> = approximateMode(classCounts, nTrain, this.rng);
      const classCountsRemaining: Type1DMatrix<number> = classCounts.map((item, index) => n_i[index] - item);
      const t_i: Type1DMatrix<number> = approximateMode(classCountsRemaining, nTest, this.rng);

      const tempTest = [];
      const tempTrain = [];

      for (let j = 0; j < nClasses; j++) {
        const permutation: Type1DMatrix<any> = this.rng.permutation(classCounts[j]);
        const permIndicesClassI = permutation.map((val) => classIndices[j][val]);
        tempTrain.push.apply(tempTrain, permIndicesClassI.slice(0, n_i[j]));
        tempTest.push.apply(tempTest, permIndicesClassI.slice(n_i[j], n_i[j] + t_i[j]));
      }
      test.push(this.rng.shuffle(tempTest));
      train.push(this.rng.shuffle(tempTrain));
    }

    return [train, test];
  };
}

function validateShuffleSplit(
  n_samples: number,
  test_size: number,
  train_size: number,
  default_test_size: number,
): Type1DMatrix<number> {
  let n_train: number;
  let n_test: number;

  if (!test_size && !train_size) {
    test_size = default_test_size;
  }

  if (test_size) {
    if (Number.isInteger(test_size)) {
      if (test_size >= n_samples || test_size <= 0) {
        throw new ValidationError(testRangeValidationError(test_size, n_samples));
      }

      n_test = test_size;
    } else {
      if (test_size <= 0 || test_size >= 1) {
        throw new ValidationError(testRangeValidationError(test_size, n_samples));
      }

      n_test = Math.ceil(test_size * n_samples);
    }
  }

  if (train_size) {
    if (Number.isInteger(train_size)) {
      if (train_size >= n_samples || train_size <= 0) {
        throw new ValidationError(trainRangeValidationError(train_size, n_samples));
      }

      n_train = train_size;
    } else {
      if (train_size <= 0 || train_size >= 1) {
        throw new ValidationError(trainRangeValidationError(train_size, n_samples));
      }

      n_train = Math.floor(train_size * n_samples);
    }
  }

  if (!train_size) {
    n_train = n_samples - n_test;
  } else if (!test_size) {
    n_test = n_samples - n_train;
  }

  const total = n_train + n_test;
  if (total > n_samples) {
    throw new ValidationError(
      `The sum of train_size and test_size = ${total}, ` +
        'should be smaller than the number of ' +
        `samples ${n_samples}. Reduce test_size and/or ` +
        'train_size.',
    );
  }

  if (n_train === 0) {
    throw new ValidationError(
      `With n_samples=${n_samples}, test_size=${test_size} and train_size=${train_size}, the ` +
        'resulting train set will be empty. Adjust any of the ' +
        'aforementioned parameters.',
    );
  }
  return [Math.floor(n_test), Math.floor(n_train)];
}
