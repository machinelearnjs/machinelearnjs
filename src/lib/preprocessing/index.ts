import { add_dummy_feature, Binarizer, MinMaxScaler, normalize, OneHotEncoder, PolynomialFeatures } from './data';
import { Imputer } from './Imputer';
import { LabelEncoder } from './label';

export {
  add_dummy_feature,
  Binarizer,
  MinMaxScaler,
  normalize,
  OneHotEncoder,
  PolynomialFeatures,
  Imputer,
  LabelEncoder,
};
