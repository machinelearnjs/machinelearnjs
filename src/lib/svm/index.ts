import { BaseSVM, NuSVC, NuSVR, OneClassSVM, SVC, SVMOptions, SVR } from './classes';

export { BaseSVM, NuSVC, NuSVR, OneClassSVM, SVC, SVMOptions, SVR };
