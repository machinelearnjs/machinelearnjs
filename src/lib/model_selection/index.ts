import { KFold, StratifiedShuffleSplit, train_test_split } from './_split';

export { KFold, train_test_split, StratifiedShuffleSplit };
