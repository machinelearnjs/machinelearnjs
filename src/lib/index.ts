import * as cluster from './cluster';
import * as datasets from './datasets';
import * as decomposition from './decomposition';
import * as ensemble from './ensemble';
import * as feature_extraction from './feature_extraction';
import * as linear_model from './linear_model';
import * as metrics from './metrics';
import * as model_selection from './model_selection';
import * as naive_bayes from './naive_bayes';
import * as neighbors from './neighbors';
import * as preprocessing from './preprocessing';
import * as svm from './svm';
import * as tree from './tree';
import * as types from './types';

// prettier-ignore
export {
	cluster,
	datasets,
	decomposition,
	ensemble,
	naive_bayes,
	feature_extraction,
	linear_model,
	metrics,
	model_selection,
	neighbors,
	preprocessing,
	svm,
	tree,
	types,
}
