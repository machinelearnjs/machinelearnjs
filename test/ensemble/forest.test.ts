import fakeFetch from 'jest-fetch-mock';
import { Iris } from '../../src/lib/datasets';
import { RandomForestClassifier } from '../../src/lib/ensemble';
import { accuracyScore } from '../../src/lib/metrics';
import { train_test_split } from '../../src/lib/model_selection';
import { matchExceptionWithSnapshot } from '../util_testing';

// Mock fetch
global.fetch = fakeFetch;

describe('ensemble:forest', () => {
  const accuracyTarget = 0.5;
  const X1 = [[0, 0], [1, 1], [2, 1], [1, 5], [3, 2]];
  const y1 = [0, 1, 2, 3, 7];
  describe('RandomForest', () => {
    it('should return a result of X1 and y1 with random state 1', () => {
      const randomForest = new RandomForestClassifier({ random_state: 1 });
      randomForest.fit(X1, y1);
      const result = randomForest.predict([[0, 3], [2, 1]]);
      const expected = [1, 2];
      expect(result).toEqual(expected);
    });

    it('should return a result of X1 and y1 with random state 2', () => {
      const randomForest = new RandomForestClassifier({ random_state: 2 });
      randomForest.fit(X1, y1);
      const result = randomForest.predict([[0, 3], [2, 1]]);
      const expected = [0, 2];
      expect(result).toEqual(expected);
    });

    it('should train test split and predict correct test #1', async () => {
      const iris = new Iris();
      const { data, targets } = await iris.load();
      const randomForest = new RandomForestClassifier();
      const { xTest, xTrain, yTest, yTrain } = train_test_split(data, targets);
      randomForest.fit(xTrain, yTrain);
      const yPred = randomForest.predict(xTest);
      const accuracy = accuracyScore(yTest, yPred);
      expect(accuracy).toBeGreaterThanOrEqual(accuracyTarget);
    });

    it('should throw an exception if incorrect inputs for fit function are given', () => {
      const randomForest = new RandomForestClassifier();
      matchExceptionWithSnapshot(randomForest.fit, [[], []]);
      matchExceptionWithSnapshot(randomForest.fit, [null, null]);
      matchExceptionWithSnapshot(randomForest.fit, [[1, 2], [3, 4]]);
      matchExceptionWithSnapshot(randomForest.fit, [[[1, 2], [3, 4]], []]);
      matchExceptionWithSnapshot(randomForest.fit, [[[1, 2], [3, 4]], [1]]);
    });

    it('should throw an exception if incorrect inputs for predict function are given', () => {
      const randomForest = new RandomForestClassifier();
      matchExceptionWithSnapshot(randomForest.predict, [[], []]);
      matchExceptionWithSnapshot(randomForest.predict, [null, null]);
      matchExceptionWithSnapshot(randomForest.predict, [[1, 2], [3, 4]]);
      matchExceptionWithSnapshot(randomForest.predict, [[[1, 2], [3, 4]], []]);
      matchExceptionWithSnapshot(randomForest.predict, [[[1, 2], [3, 4]], [1]]);
    });
  });
});
