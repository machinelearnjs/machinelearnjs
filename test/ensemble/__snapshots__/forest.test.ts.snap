// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ensemble:forest RandomForest should throw an exception if incorrect inputs for fit function are given 1`] = `[TypeError: Cannot read properties of undefined (reading 'nEstimator')]`;

exports[`ensemble:forest RandomForest should throw an exception if incorrect inputs for fit function are given 2`] = `
ValidationError {
  "message": "validateFitInputs received a non-array input X",
  "name": "ValidationError",
}
`;

exports[`ensemble:forest RandomForest should throw an exception if incorrect inputs for fit function are given 3`] = `[TypeError: Cannot read properties of undefined (reading 'nEstimator')]`;

exports[`ensemble:forest RandomForest should throw an exception if incorrect inputs for fit function are given 4`] = `
ValidationClassMismatch {
  "message": "Number of labels=0 does not math number of samples=2",
  "name": "ValidationClassMismatch",
}
`;

exports[`ensemble:forest RandomForest should throw an exception if incorrect inputs for fit function are given 5`] = `
ValidationClassMismatch {
  "message": "Number of labels=1 does not math number of samples=2",
  "name": "ValidationClassMismatch",
}
`;

exports[`ensemble:forest RandomForest should throw an exception if incorrect inputs for predict function are given 1`] = `
Validation2DMatrixError {
  "message": "The matrix is not 2D shaped: [] of [0]",
  "name": "Validation2DMatrixError",
}
`;

exports[`ensemble:forest RandomForest should throw an exception if incorrect inputs for predict function are given 2`] = `
ValidationError {
  "message": "validateMatrix2D has received a non-array argument",
  "name": "ValidationError",
}
`;

exports[`ensemble:forest RandomForest should throw an exception if incorrect inputs for predict function are given 3`] = `
Validation2DMatrixError {
  "message": "The matrix is not 2D shaped: [1,2] of [2]",
  "name": "Validation2DMatrixError",
}
`;

exports[`ensemble:forest RandomForest should throw an exception if incorrect inputs for predict function are given 4`] = `[TypeError: Cannot read properties of undefined (reading 'trees')]`;

exports[`ensemble:forest RandomForest should throw an exception if incorrect inputs for predict function are given 5`] = `[TypeError: Cannot read properties of undefined (reading 'trees')]`;
